04 - Functional Programming/04 - Functional Programming.ipynb

```diff
@@ -0,0 +1,3551 @@
+        "print(my_reduce_with_start(multiply, numbers, 1))"
+      ],
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "zJW_47dcO4NF",
+        "outputId": "e4b0e91f-3226-4dfa-a475-66a5528860e9"
+      },
+      "execution_count": 54,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "name": "stdout",
+          "text": [
+            "120\n"
+          ]
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Ето че стигнахме и до `reduce` - функция, която приема функция и колекция. Функцията бива приложена върху резултата, който имаме до момента и всеки елемент от колекцията.\n",
+        "Важно е да уточним терминологията тук - в някои езици може да срещенете тази функция като `accumulate`, `fold` или пък `foldl`.\n",
+        "\n",
+        "Тук е важно да спомена за наличието на т.нар \"ляв\" и \"десен\" `reduce`. В други функционални езици (като Haskell), имаме концепцията за ляв и десен `fold` - в каква посока се изпълнява функцията - отляво надясно или отдясно наляво. За някои операции (например умножение) реда на изпълнение няма значение, но например при деление би имало значение дали започваме от ляво или от дясно. Повече информация за fold функцията в другите езици, може да намерите [тук](https://www.wikiwand.com/en/Fold_(higher-order_function)). "
+      ],
+      "metadata": {
+        "id": "3yKqd5GiO-4b"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Вградената функция reduce в Python се намира в библиотеката `functools`. За да я използваме, първо трябва да я заредим (`import`)"
+      ],
+      "metadata": {
+        "id": "cWgQ-b6rgPBs"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "from functools import reduce\n",
+        "\n",
+        "numbers = [3, 5, 8]\n",
+        "\n",
+        "def multiply(result_so_far, current_number):\n",
+        "    return result_so_far * current_number\n",
+        "\n",
+        "print(reduce(multiply, numbers, 1))"
+      ],
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "E0OsOOpofqWV",
+        "outputId": "cb1da668-8b44-499e-f547-01ec77f6ed0f"
+      },
+      "execution_count": 55,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "name": "stdout",
+          "text": [
+            "120\n"
+          ]
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Както може би сте се досетили, тук не получаваме генератор - резултатът от `reduce` е една стойност (в нашия случай това е число).\n",
+        "\n",
+        "`reduce`, също както `map` и `filter` работи и с `lambda` функции. Тук е добро място да се отбележи, че първия аргумент, който бива подаден на функцията е стойността до момента (или стойността, която се е **акумулирала** (**accumulate**). "
```

Comment:
This is a strange way to word it. You can pass any function to it, not necessarily a lambda.

Reply:
Yeah, but the above examples are with just plain functions - I think we start the explanation with passing of normal functions, and here we expand with lambdas.